src/Features/Core/Portable/Completion/Providers/TypeImportCompletionService.cs

```diff
@@ -0,0 +1,349 @@
+
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            var cacheService = workspaceServices.GetService<IWorkspaceCacheService>();
+            if (cacheService != null)
+            {
+                cacheService.CacheFlushRequested += OnCacheFlushRequested;
+            }
+
+            return new Service(_peItemsCache, _projectItemsCache);
+        }
+
+        private void OnCacheFlushRequested(object sender, EventArgs e)
+        {
+            _peItemsCache.Clear();
+            _projectItemsCache.Clear();
+        }
+
+        private class Service : ITypeImportCompletionService
+        {
+            private readonly ConcurrentDictionary<string, ReferenceCacheEntry> _peItemsCache;
+            private readonly ConcurrentDictionary<ProjectId, ReferenceCacheEntry> _projectItemsCache;
+
+            public Service(ConcurrentDictionary<string, ReferenceCacheEntry> peReferenceCache, ConcurrentDictionary<ProjectId, ReferenceCacheEntry> projectReferenceCache)
+            {
+                _peItemsCache = peReferenceCache;
+                _projectItemsCache = projectReferenceCache;
+            }
+
+            public async Task GetAccessibleTopLevelTypesFromProjectAsync(
+                Project project,
+                Action<CompletionItem> handleAccessibleItem,
+                CancellationToken cancellationToken)
+            {
+                if (!project.SupportsCompilation)
+                {
+                    throw new ArgumentException(nameof(project));
+                }
+
+                var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                var checksum = await SymbolTreeInfo.GetSourceSymbolsChecksumAsync(project, cancellationToken).ConfigureAwait(false);
+
+                GetAccessibleTopLevelTypesWorker(
+                    project.Id,
+                    compilation.Assembly.GlobalNamespace,
+                    checksum,
+                    isInternalsVisible: true,
+                    handleAccessibleItem,
+                    _projectItemsCache,
+                    cancellationToken);
+            }
+
+            public async Task GetAccessibleTopLevelTypesFromCompilationReferenceAsync(
+                Solution solution,
+                Compilation compilation,
+                CompilationReference compilationReference,
+                Action<CompletionItem> handleAccessibleItem,
+                CancellationToken cancellationToken)
+            {
+                if (!(compilation.GetAssemblyOrModuleSymbol(compilationReference) is IAssemblySymbol assemblySymbol))
+                {
+                    return;
+                }
+
+                var isInternalsVisible = compilation.Assembly.IsSameAssemblyOrHasFriendAccessTo(assemblySymbol);
+                var assemblyProject = solution.GetProject(assemblySymbol, cancellationToken);
```

Comment:
i would check that your assemblyProject is not null

Reply:
it was originally from a compilation reference, how can it be null after a trip through workspace?

Reply:
i believe this can happen through some retargetting situations.  namely:

CompilationA references CompilationB.  

CompilationA's IAssemblySymbol for the CompilationB reference is *retargetted*.  CompilationB has it's own IAssemblySymbol.  Trying to find the project assocaited with the retargetted IAssemblySymbol will fail since it's not he same IAssemblySymbol.

Reply:
I'm wondering what should be done in this case? Maybe use references from proejct instead of Compilation?

Reply:
Yes.  that would be my definite advice.

Reply:
I have switched to using `Project.ProjectReferences` and `Project.MetadataReferences` instead, hope that would address this.

Reply:
Fixed.